lib/puppet/network/http/factory.rb

```diff
@@ -38,6 +38,11 @@ def create_connection(site)
     http.read_timeout = Puppet[:http_read_timeout]
     http.open_timeout = Puppet[:http_connect_timeout]
 
+    if Puppet[:sourceaddress]
+      Puppet.debug("Using source IP #{Puppet[:sourceaddress]}")
+      http.local_host = Puppet[:sourceaddress]
```

Comment:
Puppet still needs to support MRI 1.9.3 because it is embedded in JRuby 1.7. I would add a check here to raise an exception when running on < 2.0, something like

```ruby
if RUBY_VERSION =~ /1\./
  raise ...
else
  http.local_host = ...
end
```

Or just check if `http.responds_to?(:local_host=)`